File: clients/windows/src/DictationClient.cs
using System.IO;
using System.Net.WebSockets;
using System.Text;
using System.Text.Json;

namespace Blurt;

/// WebSocket client to the Parakeet server (see docs/protocol.md), the twin of
/// the Mac client's DictationClient.swift. Sends {start} with a fresh dictation
/// id and the declared audio format, streams PCM16 binary frames, sends
/// {stop}; surfaces info/vad/partial/final/status/error back to the caller.
/// Messages tagged with a stale dictation id (a previous dictation's late
/// final, for example) are dropped. Callbacks fire
/// on a background thread — the app marshals them onto the UI thread.
internal sealed class DictationClient
{
    /// How long the server gets to say hello before we call it unreachable.
    ///
    /// The protocol has the server send `info` the instant the socket opens, so
    /// a live blurtd answers in milliseconds. A dead one, though, does not
    /// reliably *refuse*: a host asleep behind a VPN, a port forwarded to
    /// nothing, a daemon still coming up — all of those swallow the connection
    /// and then say nothing, and ConnectAsync will wait far longer than anyone
    /// keeps talking. That wait used to pass with the mic live and the HUD
    /// reading "Listening…", so the dictation went into a socket that was never
    /// going to answer. Generous enough for a slow link, short enough that
    /// nobody gets through a sentence first. (Mirrors the Swift twin.)
    private static readonly TimeSpan HandshakeTimeout = TimeSpan.FromSeconds(3);

    public Action<string, string>? OnPartial; // (committed, live) — live may still be revised
    public Action<string>? OnFinal;
    public Action<bool>? OnVad;               // server-side VAD: is it hearing speech?
    public Action<string, string>? OnInfo;    // (state "ready|loading", model)
    public Action<string, string?>? OnStatus; // (state, detail)
    public Action<string>? OnError;
    /// The server's first message — the only proof that dictation can actually
    /// happen. Fires once per connection, ahead of the message's own callback.
    public Action? OnConnected;
    /// The server never got as far as speaking: nothing is listening, the host
    /// is unreachable, or it accepted the connection and went quiet. Kept apart
    /// from <see cref="OnError"/> because the cause is specific and the fix is
    /// actionable — start blurtd — where a mid-dictation drop is neither.
    public Action<string>? OnUnreachable;

    /// Why the last handshake was refused, when it was refused over the server's
    /// certificate. A rejected TLS handshake surfaces to <see cref="OnError"/> as
    /// an opaque exception, so the app reads this to say what actually happened —
    /// and to offer to pin the new certificate without a second handshake.
    /// Written on the TLS callback's thread, read on the UI thread.
    public volatile TrustDecision? CertRejection;

    private ClientWebSocket? _socket;
    private CancellationTokenSource? _cts;
    private volatile bool _closing;
    /// Whether anything has come back from the server on this connection. Until
    /// it has, we have no evidence of a blurtd on the other end — TCP and TLS
    /// both succeed against plenty of things that will never transcribe
    /// anything.
    private volatile bool _serverSpoke;
    /// 0/1 via Interlocked: a dying connection can fail several ways at once —
    /// the receive loop, the handshake watchdog, every queued send — and the
    /// user needs to hear about it once.
    private int _reportedFailure;
    // Written on the UI thread by ConnectAndStart, read on the receive-loop
    // thread by Handle — volatile so the reader can't see a stale value.
    private volatile string _dictationId = "";

    /// True once Close()/Cancel has begun tearing the session down. A partial can
    /// already be queued on the UI thread when that happens (e.g. Esc-cancel):
    /// delivering it would resurrect the HUD after the "Cancelled" flash and leave
    /// it stuck on screen, so callers skip late partials while this is set (mirrors
    /// the Swift `guard !self.closing` in DictationClient.swift).
    public bool Closing => _closing;

    // ClientWebSocket forbids concurrent sends; audio frames (every ~100 ms) and
    // the start/stop control messages all funnel through this gate.
    private readonly SemaphoreSlim _sendGate = new(1, 1);

    public void ConnectAndStart()
    {
        if (!Uri.TryCreate(BuildUrl(), UriKind.Absolute, out var url))
        {
            OnError?.Invoke("Bad server URL");
            return;
        }

        _closing = false;
        _serverSpoke = false;
        _reportedFailure = 0;
        _dictationId = Guid.NewGuid().ToString("N");
        _cts = new CancellationTokenSource();
        CertRejection = null;
        var socket = new ClientWebSocket();
        // Validate the server's certificate — silently. A certificate that is
        // system-valid or matches the pin for this host is accepted; anything else
        // drops the connection and records why.
        //
        // Deliberately never prompts: by the time this runs the HUD is up and the
        // mic is live, so a trust dialog here would eat the user's first sentence.
        // CertTrust.PreflightAsync settles trust before recording ever starts.
        var host = url.Host;
        var port = CertTrust.Port(url);
        socket.Options.RemoteCertificateValidationCallback = (_, cert, _, errors) =>
        {
            var decision = CertTrust.Evaluate(cert, host, port, errors);
            if (decision.Kind == TrustKind.Trusted)
            {
                CertTrust.MarkVerified(decision.Key);
                return true;
            }
            CertRejection = decision;
            // So the next dictation settles this up front rather than bringing
            // the mic up and failing the same way again.
            CertTrust.MarkRejected(decision.Key);
            return false;
        };
        _socket = socket;

        _ = RunAsync(socket, url, _cts.Token);
        _ = WatchdogAsync(_cts.Token);
    }

    /// Give up on a server that took the connection — or never even refused it —
    /// and then said nothing. A no-op once its first message has landed.
    private async Task WatchdogAsync(CancellationToken ct)
    {
        try { await Task.Delay(HandshakeTimeout, ct).ConfigureAwait(false); }
        catch (OperationCanceledException) { return; }
        if (_serverSpoke) return;
        ReportFailure($"No reply within {(int)HandshakeTimeout.TotalSeconds} seconds.");
    }

    /// Report a dead connection once, tear the socket down, and route it by
    /// whether the server ever spoke: silence from the start is an unreachable
    /// server, a drop after it spoke is a lost connection.
    private void ReportFailure(string message)
    {
        // A deliberate Close() fails everything in flight; that's not an error.
        if (_closing) return;
        if (Interlocked.Exchange(ref _reportedFailure, 1) != 0) return;
        var spoke = _serverSpoke;
        Close();
        if (spoke) OnError?.Invoke(message);
        else OnUnreachable?.Invoke(message);
    }

    private static string BuildUrl()
    {
        var url = Settings.ServerUrl.Trim();
        var token = Settings.AuthToken.Trim();
        if (token.Length == 0) return url;
        var sep = url.Contains('?') ? '&' : '?';
        return $"{url}{sep}token={Uri.EscapeDataString(token)}";
    }

    private async Task RunAsync(ClientWebSocket socket, Uri url, CancellationToken ct)
    {
        try
        {
            await socket.ConnectAsync(url, ct).ConfigureAwait(false);
            await SendJsonAsync(new
            {
                type = "start",
                id = _dictationId,
                audio = new { rate = 16000, width = 2, channels = 1 },
            }, ct).ConfigureAwait(false);
            await ReceiveLoopAsync(socket, ct).ConfigureAwait(false);
        }
        catch (Exception ex)
        {
            // A deliberate Close() cancels the socket, which surfaces here — don't
            // report that as an error (ReportFailure's `closing` guard).
            ReportFailure(ex.Message);
        }
    }

    /// Fire-and-forget send of one PCM16 frame. Copies the slice because the
    /// caller's buffer is reused by NAudio on the next callback.
    public void SendAudio(byte[] data, int count)
    {
        var socket = _socket;
        if (socket is null || socket.State != WebSocketState.Open) return;
        var frame = new byte[count];
        Buffer.BlockCopy(data, 0, frame, 0, count);
        _ = SendBinaryAsync(frame);
    }

    private async Task SendBinaryAsync(byte[] frame)
    {
        var socket = _socket;
        var ct = _cts?.Token ?? CancellationToken.None;
        if (socket is null) return;
        await _sendGate.WaitAsync(ct).ConfigureAwait(false);
        string? failure = null;
        try
        {
            if (socket.State == WebSocketState.Open)
                await socket.SendAsync(frame, WebSocketMessageType.Binary, true, ct).ConfigureAwait(false);
        }
        catch (Exception ex) { failure = ex.Message; }
        finally { _sendGate.Release(); }
        // Reported outside the gate: ReportFailure ends up marshalling onto the
        // UI thread, and nothing that slow belongs inside the send lock.
        if (failure is not null) ReportFailure(failure);
    }

    /// Ask the server to finalize; it replies with a {final} message.
    public void Stop() => _ = SendJsonAsync(new { type = "stop", id = _dictationId },
        _cts?.Token ?? CancellationToken.None);

    public void Close()
    {
        _closing = true;
        _cts?.Cancel();
        try { _socket?.Abort(); } catch { /* ignore */ }
        _socket?.Dispose();
        _socket = null;
    }

    private async Task SendJsonAsync(object obj, CancellationToken ct)
    {
        var socket = _socket;
        if (socket is null) return;
        var bytes = Encoding.UTF8.GetBytes(JsonSerializer.Serialize(obj));
        await _sendGate.WaitAsync(ct).ConfigureAwait(false);
        string? failure = null;
        try
        {
            if (socket.State == WebSocketState.Open)
                await socket.SendAsync(bytes, WebSocketMessageType.Text, true, ct).ConfigureAwait(false);
        }
        catch (Exception ex) { failure = ex.Message; }
        finally { _sendGate.Release(); }
        if (failure is not null) ReportFailure(failure);
    }

    private async Task ReceiveLoopAsync(ClientWebSocket socket, CancellationToken ct)
    {
        var buffer = new byte[8192];
        var message = new MemoryStream();
        while (socket.State == WebSocketState.Open && !ct.IsCancellationRequested)
        {
            WebSocketReceiveResult result;
            try
            {
                // ArraySegment overload → Task<WebSocketReceiveResult>; the bare
                // Memory<byte> overload would instead return a ValueWebSocketReceiveResult.
                result = await socket.ReceiveAsync(new ArraySegment<byte>(buffer), ct).ConfigureAwait(false);
            }
            catch (Exception ex)
            {
                // Once the server has spoken, WebSocketException's own wording adds
                // nothing the user can act on; before it has, the cause matters.
                ReportFailure(_serverSpoke ? "Connection lost" : ex.Message);
                return;
            }

            if (result.MessageType == WebSocketMessageType.Close) return;

            message.Write(buffer, 0, result.Count);
            if (!result.EndOfMessage) continue;

            if (result.MessageType == WebSocketMessageType.Text)
                Handle(Encoding.UTF8.GetString(message.ToArray()));
            message.SetLength(0);
        }
    }

    private void Handle(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            var root = doc.RootElement;
            if (!root.TryGetProperty("type", out var typeEl)) return;
            // Anything at all from the server proves a blurtd is on the other
            // end. Noted before the id filter below, since even a message we go
            // on to drop is evidence the connection is alive.
            if (!_serverSpoke)
            {
                _serverSpoke = true;
                OnConnected?.Invoke();
            }
            // Drop messages from a dictation that isn't ours (a late final from
            // a previous session would otherwise get typed into the wrong
            // context). Connection-scoped messages like info carry no id.
            if (root.TryGetProperty("id", out var idEl)
                && idEl.ValueKind == JsonValueKind.String
                && idEl.GetString() is { Length: > 0 } id
                && id != _dictationId)
                return;
            var text = root.TryGetProperty("text", out var t) ? t.GetString() ?? "" : "";
            switch (typeEl.GetString())
            {
                case "partial":
                    OnPartial?.Invoke(
                        root.TryGetProperty("committed", out var c) ? c.GetString() ?? "" : "",
                        root.TryGetProperty("live", out var l) ? l.GetString() ?? "" : "");
                    break;
                case "final": OnFinal?.Invoke(text); break;
                case "vad":
                    OnVad?.Invoke(root.TryGetProperty("speech", out var sp) && sp.GetBoolean());
                    break;
                case "info":
                    OnInfo?.Invoke(
                        root.TryGetProperty("state", out var st) ? st.GetString() ?? "" : "",
                        root.TryGetProperty("model", out var mo) ? mo.GetString() ?? "" : "");
                    break;
                case "status":
                    OnStatus?.Invoke(
                        root.TryGetProperty("state", out var s) ? s.GetString() ?? "" : "",
                        root.TryGetProperty("detail", out var d) ? d.GetString() : null);
                    break;
                // unknown types: ignored (forward compatibility)
            }
        }
        catch { /* malformed frame — ignore, matching the Swift best-effort parse */ }
    }
}

File: clients/mac/Sources/Blurt/DictationClient.swift
import AVFoundation
import Foundation

/// WebSocket client to the Parakeet server (see docs/protocol.md). Sends
/// {start} with a fresh dictation id and the declared audio format, streams
/// PCM16, sends {stop}; surfaces info/vad/partial/final/status back on the
/// main queue. Messages tagged with a stale dictation id (a previous
/// dictation's late final, for example) are dropped.
final class DictationClient: NSObject, URLSessionDelegate, TranscriptionEngine {
    /// How long the server gets to say hello before we call it unreachable.
    ///
    /// The protocol has the server send `info` the instant the socket opens, so
    /// a live blurtd answers in milliseconds. A dead one, though, does not
    /// reliably *refuse*: a host asleep behind a VPN, a port forwarded to
    /// nothing, a daemon still coming up — all of those swallow the connection
    /// and then say nothing, and URLSession waits a full minute before
    /// admitting it. That minute used to pass with the mic live and the HUD
    /// reading "Listening…", so the dictation went into a socket that was never
    /// going to answer. Generous enough for a slow link, short enough that
    /// nobody gets through a sentence first.
    private static let handshakeTimeout: TimeInterval = 3

    /// How long past the server's own finalization budget to wait for a `final`
    /// before giving up on it. The server promises exactly one `final` per
    /// `stop`, degrading to the committed text rather than going quiet, so
    /// overshooting its budget means something wedged rather than something
    /// slow. The margin covers the network and a server busy enough to be late
    /// answering its own deadline.
    private static let finalizeGrace: TimeInterval = 15

    /// The server's finalization budget, from `info`. The default matches
    /// blurtd's own `STOP_TIMEOUT_S` default, for a server too old to advertise
    /// it — guessing low would cut off a legitimately slow decode.
    private var stopTimeout: TimeInterval = 60

    private var task: URLSessionWebSocketTask?
    private var closing = false
    private var dictationID = ""
    private lazy var session: URLSession =
        URLSession(configuration: .default, delegate: self, delegateQueue: nil)

    /// Whether anything has come back from the server on this connection. Until
    /// it has, we have no evidence of a blurtd on the other end — TCP and TLS
    /// both succeed against plenty of things that will never transcribe
    /// anything. Main queue only.
    private var serverSpoke = false
    /// Fires once the handshake budget runs out with the server still silent.
    private var watchdog: DispatchWorkItem?
    /// Fires once a `stop` has gone unanswered past the server's own budget.
    private var finalizeWatchdog: DispatchWorkItem?
    /// A dying connection can fail several ways at once (the receive loop and
    /// every queued send). The user needs to hear about it once.
    private var reportedFailure = false

    var onPartial: ((String, String) -> Void)?   // (committed, live) — live may still be revised
    var onFinal: ((String) -> Void)?
    var onVad: ((Bool) -> Void)?                 // server-side VAD: is it hearing speech?
    var onInfo: ((String, String) -> Void)?      // (state "ready|loading", model)
    var onStatus: ((String, String?) -> Void)?   // (state, detail)
    var onError: ((String) -> Void)?
    /// The server's first message — the only proof that dictation can actually
    /// happen. Fires once per connection, ahead of the message's own callback.
    var onConnected: (() -> Void)?
    /// The server never got as far as speaking: nothing is listening, the host
    /// is unreachable, or it accepted the connection and went quiet. Kept apart
    /// from `onError` because the cause is specific and the fix is actionable —
    /// start blurtd — where a mid-dictation drop is neither.
    var onUnreachable: ((String) -> Void)?
    /// `stop` was acknowledged by nothing: the socket is still up but the
    /// server's `final` never came, and waiting longer won't help.
    var onFinalizeTimeout: (() -> Void)?

    /// Why the last handshake was refused, when it was refused over the server's
    /// certificate. A rejected TLS handshake surfaces to `onError` as an opaque
    /// "connection lost", so the app reads this to say what actually happened —
    /// and to offer to pin the new certificate without a second handshake.
    /// Main queue only.
    private(set) var certRejection: (decision: CertTrust.Decision, host: String, port: Int)?

    func connectAndStart() {
        // Cleared before the guards below, so a stale rejection can't make a
        // plain bad-URL error come back as a certificate dialog.
        certRejection = nil
        guard var comps = URLComponents(string: Settings.serverURL) else {
            onError?("Bad server URL"); return
        }
        if !Settings.authToken.isEmpty {
            comps.queryItems = [URLQueryItem(name: "token", value: Settings.authToken)]
        }
        guard let url = comps.url else { onError?("Bad server URL"); return }
        closing = false
        serverSpoke = false
        reportedFailure = false
        finalizeWatchdog?.cancel()   // defensive: a previous session's wait, if any
        finalizeWatchdog = nil
        dictationID = UUID().uuidString
        let t = session.webSocketTask(with: url)
        task = t
        t.resume()
        sendJSON(["type": "start", "id": dictationID,
                  "audio": ["rate": 16000, "width": 2, "channels": 1]])
        receiveLoop()
        armWatchdog()
    }

    /// The wire format the server declares in `start` — the same one
    /// AudioCapture is told to produce, so this is a straight reinterpret of the
    /// buffer's bytes rather than a conversion.
    var inputFormat: AVAudioFormat { AudioCapture.serverFormat }

    func sendAudio(_ buffer: AVAudioPCMBuffer) {
        guard let ch = buffer.int16ChannelData, buffer.frameLength > 0 else { return }
        let data = Data(bytes: ch[0], count: Int(buffer.frameLength) * MemoryLayout<Int16>.size)
        task?.send(.data(data)) { [weak self] error in
            guard let error else { return }
            DispatchQueue.main.async { self?.reportFailure(error.localizedDescription) }
        }
    }

    /// Ask the server to finalize; it will reply with a {final} message.
    func stop() {
        sendJSON(["type": "stop", "id": dictationID])
        armFinalizeWatchdog()
    }

    func close() {
        closing = true
        watchdog?.cancel()
        watchdog = nil
        finalizeWatchdog?.cancel()
        finalizeWatchdog = nil
        task?.cancel(with: .goingAway, reason: nil)
        task = nil
    }

    private func sendJSON(_ obj: [String: Any]) {
        guard let d = try? JSONSerialization.data(withJSONObject: obj),
              let s = String(data: d, encoding: .utf8) else { return }
        task?.send(.string(s)) { [weak self] error in
            guard let error else { return }
            DispatchQueue.main.async { self?.reportFailure(error.localizedDescription) }
        }
    }

    /// Give up on a server that took the connection and then said nothing.
    /// Cancelled by its first message, and by `close()`. Main queue only.
    private func armWatchdog() {
        watchdog?.cancel()
        let work = DispatchWorkItem { [weak self] in
            guard let self, !self.serverSpoke else { return }
            self.reportFailure("No reply within \(Int(Self.handshakeTimeout)) seconds.")
        }
        watchdog = work
        DispatchQueue.main.asyncAfter(deadline: .now() + Self.handshakeTimeout, execute: work)
    }

    /// Stop waiting on a `final` that isn't coming. A wedged server (or one
    /// killed between the `stop` and the reply) leaves an otherwise healthy
    /// socket with nothing on it, which the receive loop has no way to notice —
    /// so without this the HUD sits on the last partial for good. Cancelled by
    /// the `final` itself, and by `close()`. Main queue only.
    private func armFinalizeWatchdog() {
        finalizeWatchdog?.cancel()
        let work = DispatchWorkItem { [weak self] in
            guard let self, !self.closing else { return }
            self.close()
            self.onFinalizeTimeout?()
        }
        finalizeWatchdog = work
        DispatchQueue.main.asyncAfter(deadline: .now() + stopTimeout + Self.finalizeGrace,
                                      execute: work)
    }

    /// Report a dead connection once, tear the socket down, and route it by
    /// whether the server ever spoke: silence from the start is an unreachable
    /// server, a drop after it spoke is a lost connection. Main queue only.
    private func reportFailure(_ message: String) {
        guard !reportedFailure, !closing else { return }
        reportedFailure = true
        let spoke = serverSpoke
        close()
        if spoke { onError?(message) } else { onUnreachable?(message) }
    }

    private func receiveLoop() {
        task?.receive { [weak self] result in
            guard let self = self else { return }
            switch result {
            case .failure(let err):
                // A deliberate close() cancels the socket, which surfaces here as
                // a "socket is not connected" failure. Don't report that as an error.
                if self.closing { return }
                DispatchQueue.main.async { self.reportFailure(err.localizedDescription) }
            case .success(let message):
                if case .string(let s) = message { self.handle(s) }
                self.receiveLoop()
            }
        }
    }

    private func handle(_ s: String) {
        guard let d = s.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: d) as? [String: Any],
              let type = obj["type"] as? String else { return }
        DispatchQueue.main.async {
            // A partial can already be queued on main when close() is called
            // (e.g. Esc-cancel): delivering it would resurrect the HUD after
            // the "Cancelled" flash and leave it stuck on screen.
            guard !self.closing else { return }
            // Anything at all from the server proves a blurtd is on the other
            // end. Noted before the id filter below, since even a message we go
            // on to drop is evidence the connection is alive.
            if !self.serverSpoke {
                self.serverSpoke = true
                self.watchdog?.cancel()
                self.watchdog = nil
                self.onConnected?()
            }
            // Drop messages from a dictation that isn't ours (a late final from
            // a previous session would otherwise get typed into the wrong
            // context). Checked here rather than on the delegate queue so that
            // `dictationID` is only ever touched on main — this callback runs
            // on a URLSession queue, and a String assignment racing with a read
            // is not merely stale, it's undefined. Connection-scoped messages
            // like info carry no id and pass through.
            if let id = obj["id"] as? String, !id.isEmpty, id != self.dictationID { return }
            switch type {
            case "partial":
                self.onPartial?(obj["committed"] as? String ?? "",
                                obj["live"] as? String ?? "")
            case "final":
                // Cancelled here rather than leaning on the close() that
                // follows in onFinal: if the final lands exactly on the
                // deadline both are already queued, and this one is first.
                self.finalizeWatchdog?.cancel()
                self.finalizeWatchdog = nil
                self.onFinal?(obj["text"] as? String ?? "")
            case "vad":    self.onVad?(obj["speech"] as? Bool ?? false)
            case "info":
                // Additive field: a server too old to send it leaves the
                // default standing (see stopTimeout).
                if let budget = obj["stop_timeout_s"] as? Double, budget > 0 {
                    self.stopTimeout = budget
                }
                self.onInfo?(obj["state"] as? String ?? "",
                             obj["model"] as? String ?? "")
            case "status": self.onStatus?(obj["state"] as? String ?? "", obj["detail"] as? String)
            default: break   // unknown types: ignored (forward compatibility)
            }
        }
    }

    /// Validate the server's certificate — silently. A certificate that is
    /// system-valid or matches the pin for this host is accepted; anything else
    /// drops the connection and records why in `certRejection`.
    ///
    /// Deliberately never prompts: by the time this runs the HUD is up and the
    /// mic is live, so a trust dialog here would eat the user's first sentence.
    /// CertTrust.preflight settles trust before recording ever starts.
    func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
        guard let trust = challenge.protectionSpace.serverTrust else {
            completionHandler(.performDefaultHandling, nil)
            return
        }
        let host = challenge.protectionSpace.host
        let port = challenge.protectionSpace.port
        let decision = CertTrust.evaluate(trust: trust, host: host, port: port)
        if case .trusted = decision {
            DispatchQueue.main.async { CertTrust.markVerified(CertTrust.pinKey(host: host, port: port)) }
            completionHandler(.useCredential, URLCredential(trust: trust))
            return
        }
        // Enqueued before the handshake is failed, so it lands on main ahead of
        // the onError that reads it.
        DispatchQueue.main.async {
            self.certRejection = (decision, host, port)
            // So the next dictation settles this up front rather than bringing
            // the mic up and failing the same way again.
            CertTrust.markRejected(CertTrust.pinKey(host: host, port: port))
        }
        completionHandler(.cancelAuthenticationChallenge, nil)
    }
}
